plutus-scb/src/Plutus/SCB/Webserver/Handler.hs

```diff
@@ -0,0 +1,212 @@
+activateContract = Core.activateContract
+
+getContractInstanceState ::
+       forall t effs.
+       ( Member (EventLogEffect (ChainEvent t)) effs
+       , Member (Error SCBError) effs
+       )
+    => ContractInstanceId
+    -> Eff effs (ContractInstanceState t)
+getContractInstanceState contractId = do
+    contractStates <- runGlobalQuery (Query.contractState @t)
+    case Map.lookup contractId contractStates of
+        Nothing    -> throwError $ ContractInstanceNotFound contractId
+        Just value -> pure value
+
+invokeEndpoint ::
+       forall t effs.
+       ( Member (EventLogEffect (ChainEvent t)) effs
+       , Member (ContractEffect t) effs
+       , Member (LogMsg ContractExeLogMsg) effs
+       , Member (Error SCBError) effs
+       , Member (LogMsg (Instance.ContractInstanceMsg t)) effs
+       , Member (LogObserve (LogMessage Text)) effs
+       , Pretty t
+       )
+    => EndpointDescription
+    -> JSON.Value
+    -> ContractInstanceId
+    -> Eff effs (ContractInstanceState t)
+invokeEndpoint (EndpointDescription endpointDescription) payload contractId = do
+    logInfo $ InvokingEndpoint endpointDescription payload
+    newState :: [ChainEvent t] <-
+        Instance.callContractEndpoint @t contractId endpointDescription payload
+    logInfo $ EndpointInvocationResponse $ fmap pretty newState
+    getContractInstanceState contractId
+
+parseContractId ::
+       (Member (Error SCBError) effs) => Text -> Eff effs ContractInstanceId
+parseContractId t =
+    case UUID.fromText t of
+        Just uuid -> pure $ ContractInstanceId uuid
+        Nothing   -> throwError $ InvalidUUIDError t
+
+
+handler ::
+       forall effs.
+       ( Member (EventLogEffect (ChainEvent ContractExe)) effs
+       , Member (ContractEffect ContractExe) effs
+       , Member ChainIndexEffect effs
+       , Member UUIDEffect effs
+       , Member (LogMsg ContractExeLogMsg) effs
+       , Member (Error SCBError) effs
+       , Member (LogMsg UnStringifyJSONLog) effs
+       , Member (LogMsg (Instance.ContractInstanceMsg ContractExe)) effs
+       , Member (LogObserve (LogMessage Text)) effs
+       )
+    => Eff effs ()
+       :<|> (Eff effs (FullReport ContractExe)
+             :<|> ((ContractExe -> Eff effs (ContractInstanceState ContractExe))
+                   :<|> (Text -> Eff effs (ContractSignatureResponse ContractExe)
+                                 :<|> (String -> JSON.Value -> Eff effs (ContractInstanceState ContractExe)))))
+handler =
+    healthcheck :<|> getFullReport :<|>
+    (activateContract :<|>
+     (\rawInstanceId ->
+          (parseContractId rawInstanceId >>= contractSchema) :<|>
+          (\rawEndpointDescription payload ->
+               parseStringifiedJSON payload >>= \payload' ->
+                   parseContractId rawInstanceId >>=
+                   invokeEndpoint
+                       (EndpointDescription rawEndpointDescription)
+                       payload')))
```

Comment:
this whole thing is very difficult to read

Reply:
Yup. Improved. 👍